mesonbuild/compilers/detect.py

```diff
@@ -270,36 +158,38 @@ def _handle_exceptions(
 # ===============
 
 def detect_static_linker(env: 'Environment', compiler: Compiler) -> StaticLinker:
```

Comment:
lint / mypy errors out here:

```
mesonbuild/compilers/detect.py:160:1: error: Missing return statement
```

@dcbaker I have zero idea why this happens. I have not changed anything that could possibly matter here.

Reply:
@eli-schwartz every other places has `raise EnvironmentException('Unreachable code (exception to make mypy happy)')`. You return None at the very end of that function otherwise.

Reply:
I, personally, don't do anything.

This function, before I got here, ends with running a function annotated as `T.NoReturn` because it cannot ever return. The function exists *solely* to raise MesonException with specially formatted contents.

Reply:
And that is *exactly* what `NoReturn` is supposed to do, it's supposed to tell mypy "this will always raise an exception, everything past this is dead code"

Reply:
Well, to the surprise of exactly nobody, mypy is buggy and it is not new because we already have `raise EnvironmentException('Unreachable code (exception to make mypy happy)')` everywhere after that NoReturn function. TBH, just add that line and this PR looks good to me.

Reply:
I can reproduce the error `mesonbuild/compilers/detect.py:160:1: error: Missing return statement` locally with this PR, adding that line fixes it as expected.

Reply:
What burns me is that locally running mypy with the exact same version as the one installed in CI, I can't reproduce the error!

Reply:
@eli-schwartz what version of mypy? I can't reproduce with 0.961 on python 3.10.6 either you're PR passes fine.

Reply:
0.971

I discovered something amazing.
```
$ mypy mesonbuild/compilers/
Success: no issues found in 32 source files

$ ./run_mypy.py 
Running mypy (this can take some time) ...
mesonbuild/compilers/detect.py:160:1: error: Missing return statement
Found 1 error in 1 file (checked 173 source files)

$ mypy mesonbuild/compilers/
mesonbuild/compilers/detect.py:160:1: error: Missing return statement
Found 1 error in 1 file (checked 32 source files)

$ mypy mesonbuild/compilers/detect.py
Success: no issues found in 1 source file

```

I have no explanation.

Reply:
I just rebased a bit, and broke out some prep changes into their own commits for better reviewability, particularly:
>  - compilers/detect: rename potentially conflicting name
>  - compilers: use more direct checks for what kind of compiler we have

I also moved around the CsCompiler and DCompiler assignments to avoid cases where inside the middle of the patch series, mypy failed to run (in git bisect).

The mypy failure only happens for "compilers: avoid importing compilers upfront for detect" but there's less changes in that file now, so it's easier to look at.

The really, amazingly, jaw-droppingly, awe-inspiring thing?

The mypy failure goes away if I revert the beginning-of-the-file changes to the imports. I can't even.